tracing-subscriber/src/fmt/format/json.rs

```diff
@@ -89,37 +120,21 @@ where
             let id = id.id();
             if let Some(id) = id {
                 if let Some(span) = ctx.ctx.span(id) {
-                    let ext = span.extensions();
-                    let data = ext
-                        .get::<FormattedFields<N>>()
-                        .expect("Unable to find FormattedFields in extensions; this is a bug");
-                    // TODO: let's _not_ do this, but this resolves
-                    // https://github.com/tokio-rs/tracing/issues/391.
-                    // We should probably rework this to use a `serde_json::Value` or something
-                    // similar in a JSON-specific layer, but I'd (david)
-                    // rather have a uglier fix now rather than shipping broken JSON.
-                    let mut fields = match serde_json::from_str(&data) {
-                        Ok(fields) => fields,
-                        // We have previously recorded fields for this span
-                        // should be valid JSON. However, they appear to *not*
-                        // be valid JSON. This is almost certainly a bug, so
-                        // panic if we're in debug mode
-                        Err(e) if cfg!(debug_assertions) => panic!(
-                            "span '{}' had malformed fields! this is a bug.\n  error: {}\n  fields: {:?}",
-                            span.metadata().name(),
-                            e,
-                            data
-                        ),
-                        // If we *aren't* in debug mode, it's probably best not
-                        // crash the program, but let's at least make sure it's clear
-                        // that the fields are not supposed to be missing.
-                        Err(e) => json!({ "field_error": format!("{}", e) }),
-                    };
-                    fields["name"] = json!(span.metadata().name());
-                    serializer.serialize_entry("span", &fields).unwrap_or(());
+                    serializer
+                        .serialize_entry("span", &visit_span(&span))
+                        .unwrap_or(());
                 }
             }
 
+            let mut spans = vec![];
+            ctx.visit_spans(|span| {
+                spans.push(visit_span(span));
+                Ok(())
+            })?;
+            if !spans.is_empty() {
+                serializer.serialize_entry("spans", &spans)?;
+            }
```

Comment:
Rather than `collect`ing to a `Vec`, which allocates, it should be possible to serialize the spans without an intermediate allocation, using `serde`'s [`SerializeSeq`](https://docs.serde.rs/serde/ser/trait.SerializeSeq.html). This might offer a small performance improvement.

I think we could do this by adding a newtype which wraps the `Context::spans` iterator and implements `serde::Serialize` like this:

```rust
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
         let mut seq = s.serialize_seq();
         for span in self.0 {
             // ...
         }
    }
```
and passing that struct to `serialize_entry` as the value.

Of course, this would involve a little more complexity than the current approach, so I'm fine with merging what we have now and doing this optimization as a follow-up. We might want to do some benchmarking to determine how much of a performance improvement we would get from it.

Reply:
I implemented `Serialize` for newtypes for the context and the span, to keep it all managed by serde.
Please let me know if you see any issue in how I did it